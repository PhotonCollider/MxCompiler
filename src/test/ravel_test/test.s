.text
	.globl main
point.printPoint:
	addi	sp, sp, -80
	sw	ra, 76(sp)
	mv	t2, a0
	addi	t2, t2, 0
	sw	t2, 0(sp)
	lw	t0, 0(sp)
	lw	t1, 0(t0)
	sw	t1, 4(sp)
	sw	a0, 60(sp)
	lw	a0, 4(sp)
	call	toString
	sw	a0, 8(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	la	a0, stringLiteral.0
	lw	a1, 8(sp)
	call	builtin.string_add
	sw	a0, 12(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	lw	a0, 12(sp)
	la	a1, stringLiteral.1
	call	builtin.string_add
	sw	a0, 16(sp)
	lw	a0, 60(sp)
	mv	t2, a0
	addi	t2, t2, 4
	sw	t2, 20(sp)
	lw	t0, 20(sp)
	lw	t1, 0(t0)
	sw	t1, 24(sp)
	sw	a0, 60(sp)
	lw	a0, 24(sp)
	call	toString
	sw	a0, 28(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	lw	a0, 16(sp)
	lw	a1, 28(sp)
	call	builtin.string_add
	sw	a0, 32(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	lw	a0, 32(sp)
	la	a1, stringLiteral.2
	call	builtin.string_add
	sw	a0, 36(sp)
	lw	a0, 60(sp)
	mv	t2, a0
	addi	t2, t2, 8
	sw	t2, 40(sp)
	lw	t0, 40(sp)
	lw	t1, 0(t0)
	sw	t1, 44(sp)
	sw	a0, 60(sp)
	lw	a0, 44(sp)
	call	toString
	sw	a0, 48(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	lw	a0, 36(sp)
	lw	a1, 48(sp)
	call	builtin.string_add
	sw	a0, 52(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	lw	a0, 52(sp)
	la	a1, stringLiteral.3
	call	builtin.string_add
	sw	a0, 56(sp)
	lw	a0, 60(sp)
	sw	a0, 60(sp)
	lw	a0, 56(sp)
	call	println
	lw	a0, 60(sp)
	lw	ra, 76(sp)
	addi	sp, sp, 80
	ret
point.cross:
	addi	sp, sp, -32
	sw	ra, 28(sp)

	mv	t1, a1
	sw	t1, 0(sp)

	sw	a0, 20(sp)
	sw	a1, 24(sp)
	li	a0, 12
	call	builtin.calloc
	sw	a0, 8(sp)
	lw	a0, 20(sp)
	lw	a1, 24(sp)

	call	point.printPoint

	sw	a0, 20(sp)
	sw	a1, 24(sp)
	call	point.point
	lw	a0, 20(sp)
	lw	a1, 24(sp)

	call	point.printPoint

	lw	t1, 8(sp)
	sw	t1, 4(sp)
	lw	t1, 4(sp)
	sw	t1, 12(sp)

	sw	a0, 20(sp)
	sw	a1, 24(sp)
	lw	a0, 12(sp)
	li	a1, 7
	li	a2, 7
	li	a3, 7
	call	point.set
	lw	a0, 20(sp)
	lw	a1, 24(sp)

	lw	t1, 4(sp)
	sw	t1, 16(sp)
	lw	a0, 16(sp)
	lw	ra, 28(sp)
	addi	sp, sp, 32
	ret
main:
	addi	sp, sp, -64
	sw	ra, 60(sp)
	call	global.init
	li	a0, 12
	call	builtin.calloc
	sw	a0, 4(sp)
	call	point.point
	lw	t1, 4(sp)
	sw	t1, 0(sp)
	li	a0, 12
	call	builtin.calloc
	sw	a0, 12(sp)
	call	point.point
	lw	t1, 12(sp)
	sw	t1, 8(sp)
	lw	t1, 0(sp)
	sw	t1, 16(sp)
	lw	a0, 16(sp)
	li	a1, 1
	li	a2, 2
	li	a3, 3
	call	point.set
	lw	t1, 8(sp)
	sw	t1, 20(sp)
	lw	a0, 20(sp)
	li	a1, 4
	li	a2, 5
	li	a3, 6
	call	point.set
	lw	t1, 0(sp)
	sw	t1, 24(sp)
	lw	a0, 24(sp)
	call	point.printPoint
	lw	t1, 8(sp)
	sw	t1, 28(sp)
	lw	a0, 28(sp)
	call	point.printPoint

    lw  a0, 0(sp)
    call    point.printPoint

	lw	t1, 0(sp)
	sw	t1, 32(sp)
	lw	t1, 8(sp)
	sw	t1, 36(sp)

	lw	a0, 32(sp)
	lw	a1, 36(sp)
	call	point.cross
	sw	a0, 40(sp)

	lw  a0, 0(sp)
    call	point.printPoint

	li	a0, 0
	lw	ra, 60(sp)
	addi	sp, sp, 64
	ret
point.set:
	addi	sp, sp, -64
	sw	ra, 60(sp)
	mv	t1, a1
	sw	t1, 0(sp)
	mv	t1, a2
	sw	t1, 4(sp)
	mv	t1, a3
	sw	t1, 8(sp)
	lw	t1, 0(sp)
	sw	t1, 12(sp)
	mv	t2, a0
	addi	t2, t2, 0
	sw	t2, 16(sp)
	lw	t0, 16(sp)
	lw	t1, 12(sp)
	sw	t1, 0(t0)
	lw	t1, 4(sp)
	sw	t1, 20(sp)
	mv	t2, a0
	addi	t2, t2, 4
	sw	t2, 24(sp)
	lw	t0, 24(sp)
	lw	t1, 20(sp)
	sw	t1, 0(t0)
	lw	t1, 8(sp)
	sw	t1, 28(sp)
	mv	t2, a0
	addi	t2, t2, 8
	sw	t2, 32(sp)
	lw	t0, 32(sp)
	lw	t1, 28(sp)
	sw	t1, 0(t0)
	lw	ra, 60(sp)
	addi	sp, sp, 64
	ret
point.point:
	addi	sp, sp, -32
	sw	ra, 28(sp)
	mv	t2, a0
	addi	t2, t2, 0
	sw	t2, 0(sp)
	lw	t0, 0(sp)
	li	t1, 0
	sw	t1, 0(t0)
	mv	t2, a0
	addi	t2, t2, 4
	sw	t2, 4(sp)
	lw	t0, 4(sp)
	li	t1, 0
	sw	t1, 0(t0)
	mv	t2, a0
	addi	t2, t2, 8
	sw	t2, 8(sp)
	lw	t0, 8(sp)
	li	t1, 0
	sw	t1, 0(t0)
	lw	ra, 28(sp)
	addi	sp, sp, 32
	ret
global.init:
	addi	sp, sp, -16
	sw	ra, 12(sp)
	lw	ra, 12(sp)
	addi	sp, sp, 16
	ret

.data

.rodata
stringLiteral.0:
	.asciz "("
stringLiteral.1:
	.asciz ", "
stringLiteral.2:
	.asciz ", "
stringLiteral.3:
	.asciz ")"

