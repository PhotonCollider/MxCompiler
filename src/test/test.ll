;Definition Of Structs

;Definition Of Global Variables

;Definition Of String Literals

;Definition Of Formatted String Fragments

;Definition Of Array Literals

;Declarations Of Functions
declare void @print(ptr)
declare void @println(ptr)
declare void @printInt(i32)
declare void @printlnInt(i32)
declare ptr @getString()
declare i32 @getInt()
declare ptr @toString(i32)
declare ptr @array.copy(ptr, i32)
declare i32 @array.size(ptr)
declare i32 @string.length(ptr)
declare i32 @string.ord(ptr, i32)
declare i32 @string.parseInt(ptr)
declare ptr @string.substring(ptr, i32, i32)
declare ptr @builtin.bool_to_string(i1)
declare ptr @builtin.calloc(i32)
declare ptr @builtin.calloc_array(i32, i32)
declare ptr @builtin.malloc(i32)
declare ptr @builtin.malloc_array(i32, i32)
declare ptr @builtin.string_add(ptr, ptr)
declare i1 @builtin.string_eq(ptr, ptr)
declare i1 @builtin.string_ge(ptr, ptr)
declare i1 @builtin.string_geq(ptr, ptr)
declare i1 @builtin.string_le(ptr, ptr)
declare i1 @builtin.string_leq(ptr, ptr)
declare i1 @builtin.string_ne(ptr, ptr)

;Definition Of Functions
define i32 @main () {
main:
	call void @global.init()
	%N.1 = alloca i32
	store i32 5, ptr %N.1
	%i.1 = alloca i32
	store i32 2, ptr %i.1
	br label %for.cond.0
for.cond.0:
	br i1 true, label %for.body.0, label %for.end.0
for.body.0:
	%0 = load i32, ptr %i.1
	%1 = load i32, ptr %N.1
	%2 = icmp sgt i32 %0, %1
	br i1 %2, label %if.then.0, label %if.else.0
if.then.0:
	br label %for.end.0
if.else.0:
	br label %if.endif.0
if.endif.0:
	%i.1 = alloca i32
	br label %for.update.0
for.update.0:
	%3 = load i32, ptr %i.1
	%4 = add i32 %3, 1
	store i32 %4, ptr %i.1
	br label %for.cond.0
for.end.0:
	ret i32 0
}

define void @global.init () {
global.init:
	ret void
}

