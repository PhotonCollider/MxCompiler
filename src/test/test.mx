/*
Test Package: Codegen
Author: Haojun Mao
Time: 2020-01-25
Input:
=== input ===
1
2
3
4
=== end ===
Output:
=== output ===
0000
=== end ===
ExitCode: 0
InstLimit: -1
Origin Package: Codegen Pretest-531
*/
//input: 1 2 3 4

int[] a = new int[4];
int main()
{
		a[0] = 0;

	a[1] = 1;
	a[2] = 2;
	a[3] = 3;

	//toString(a[0]);
	toString(a[1]);
	println(toString(a[1]));
	println(toString(a[2]));
	println(toString(a[3]));

	for (int i = 0; i < 4; i++)
	{
		println(toString(a[i]));
	}
    return 0;
}