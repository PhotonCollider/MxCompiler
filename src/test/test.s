.text
main:
	addi sp, sp, -16
	sw ra, 12(sp)
	call global.init
	la a0, stringLiteral.0
	call println
	li a0, 0
	lw ra, 12(sp)
	addi sp, sp, 16
global.init:
	addi sp, sp, -16
	sw ra, 12(sp)
	lw ra, 12(sp)
	addi sp, sp, 16

.data

.rodata
stringLiteral.0:
	.asciz "hello, world"

