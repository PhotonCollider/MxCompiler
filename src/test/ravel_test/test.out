(849, -463, 480)
